File: Study Plan/LeetCode 75/0443 String Compression/StringCompression.java
/*  
    Problem
    --------------------------------------------------------------------------------
    443. String Compression
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 1 ms (Beats 99.39% of users with Java)
    - Memory 44.04 MB (Beats 45.58% of users with Java)
*/

// Editorial
public class Solution {
    public int compress(char[] chars) {
        int index = 0;
        int position = 0;
        while (index < chars.length) {
            int consecutive = 1;
            while (index + consecutive < chars.length
                && chars[index + consecutive] == chars[index]) {
                consecutive++;
            }
            chars[position++] = chars[index];
            if (consecutive > 1) {
                for(char character : Integer.toString(consecutive).toCharArray()) {
                    chars[position++] = character;
                }
            }
            index += consecutive;
        }
        return position;
    }
}

File: Study Plan/LeetCode 75/0443 String Compression/StringCompression.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    443. String Compression
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 127 ms (Beats 62.71% of users with C#)
    - Memory 62.90 MB (Beats 83.06% of users with C#)
*/

// Editorial
public class Solution
{
    public int Compress(char[] chars)
    {
        var index = 0;
        var position = 0;

        while (index < chars.Length)
        {
            var consecutive = 1;

            while (index + consecutive < chars.Length
                && chars[index + consecutive] == chars[index])
            {
                consecutive++;
            }

            chars[position++] = chars[index];

            if (consecutive > 1)
            {
                foreach (var character in consecutive.ToString())
                {
                    chars[position++] = character;
                }
            }

            index += consecutive;
        }

        return position;
    }
}
